### PR TITLE
Implement Hand-Picked Products block

### LINKED ISSUE #7633: Product Query [3]: [Variation] `Hand-picked Products` block
We want to implement Hand-picked products as a variation of the Product Query block. As of the time of writing, this is going the be the only variation of the Product Query block as it will work significantly differently. Likely all filters need to be disabled and a special input to select products should be shown. Also, there should be an easy way to custom sort the products if required. We probably want to enable users to have both custom and auto sorting orders options.

Design TBD. cc @vivialice @nerrad

### PR DESCRIPTION
**Pull request by sunyatasattva**
This PR implements the `ProductSelector` advanced filter within the “Products (Beta)” block. The filter allows the merchant to narrow down the exact products to which all subsequent filters will be applied, mirroring the functionality of the existing “Hand-picked Products” plus all the other functionalities available from the “Products (Beta)” block.

Fixes #7633

#### Accessibility

- [x] I've tested using only a keyboard (no mouse)
- [ ] I've tested using a screen reader
- [ ] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
- [x] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

#### Other Checks

- [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] This PR adds/removes an experimental interfaces and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] I tagged two reviewers because this PR makes queries to the database or I think it might have some security impact.

### Screenshots

https://user-images.githubusercontent.com/1847066/207167971-8b550a4d-a14b-458b-8573-9a9e02d97e3a.mov

The video above shows a different block called “Hand-picked Products”. The current PR doesn't include this variation and includes the Product Selector within the normal block.

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests

##### Simplest path

1. Add a “Products (Beta)” block to your page.
2. Go to the “Advanced Filters” within the Inspector Controls and add “Hand-picked Products”.
3. Start typing to select some products.
4. Ensure the preview is correctly updated as you narrow down your selection.
5. Click Publish.
6. Ensure the front-end is correct.

##### Complex path

1. Repeat steps 1–4 above.
2. Add additional filters (such as “On sale”).
3. Ensure the block only applies the subsequent filters to the subset of products you have selected.

* [ ] Do not include in the Testing Notes 

### WooCommerce Visibility

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Changelog

> Products (Beta): Add a filter to allow merchants to hand-pick a subset of products.

### CODE REVIEW

**Comment by kmanijak** [comment]
PR bumped to the next milestone (10.3.0)

File: assets/js/blocks/product-query/inspector-controls/product-selector.tsx [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+	__experimentalToolsPanelItem as ToolsPanelItem,
+} from '@wordpress/components';
+
+/**
+ * Internal dependencies
+ */
+import { ProductQueryBlock } from '../types';
+import { setQueryAttribute } from '../utils';
+
+export const ProductSelector = ( props: ProductQueryBlock ) => {
```
Comment (imanish003) [inline assets/js/blocks/product-query/inspector-controls/product-selector.tsx]: I am nitpicking here, but how do you feel about restructuring the attributes property from the props object to simplify property access:
```suggestion
export const ProductSelector = ({ attributes }: ProductQueryBlock) => {
```

File: assets/js/blocks/product-query/inspector-controls/product-selector.tsx [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+	const [ productsList, setProductsList ] = useState< ProductResponseItem[] >(
+		[]
+	);
+
+	useEffect( () => {
+		getProducts( { selected: [] } ).then( ( results ) => {
+			setProductsList( results as ProductResponseItem[] );
+		} );
+	}, [] );
```
Comment (imanish003) [inline assets/js/blocks/product-query/inspector-controls/product-selector.tsx:30]: I don't think it's needed here but just wanted to share my thoughts. Maybe we can create a custom hook to fetch and manage the productsList state to separate concerns and make the component more modular:
```jsx
const useProductsList = (): ProductResponseItem[] => {
  const [productsList, setProductsList] = useState<ProductResponseItem[]>([]);

  useEffect(() => {
    getProducts({ selected: [] }).then((results) => {
      setProductsList(results as ProductResponseItem[]);
    });
  }, []);

  return productsList;
};
```

And use it in the component like this:
```jsx
const productsList = useProductsList();
```

File: assets/js/blocks/product-query/inspector-controls/product-selector.tsx [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+								},
+							},
+							{}
+						);
+					} else {
+						setQueryAttribute( props, {
+							include: ids,
+						} );
+					}
+				} }
```
Comment (imanish003) [inline assets/js/blocks/product-query/inspector-controls/product-selector.tsx]: Maybe we can extract the logic of the onChange callback into a separate function to make the component more readable and maintainable?
```jsx
const handleTokensChange = (values: string[]) => {
  ...
};

...
<FormTokenField
  ...
  onChange={handleTokensChange}
  ...
/>
```

File: assets/js/blocks/product-query/constants.ts [hunk 4]
```diff
@@ -12,15 +13,6 @@ import { VARIATION_NAME as PRODUCT_TITLE_ID } from './variations/elements/produc
...
 
-/**
- * Returns an object without a key.
- */
-function objectOmit< T, K extends keyof T >( obj: T, key: K ) {
-	const { [ key ]: omit, ...rest } = obj;
-
-	return rest;
-}
-
```
Comment (imanish003) [inline assets/js/blocks/product-query/constants.ts:23]: Great! Thanks for moving this to utils 🙌🏻

**Review by imanish003 — APPROVED** [review]
Hi @sunyatasattva,
I've tested the changes, and everything is working as expected. Excellent work, as always 🙌🏻
I've left a few minor comments, but they shouldn't block the PR, so I'm pre-approving it. 🚀

File: assets/js/blocks/product-query/inspector-controls/product-selector.tsx (hunk 1 above)
Comment (sunyatasattva) [reply assets/js/blocks/product-query/inspector-controls/product-selector.tsx]: Thanks for the suggestion! Normally, I'd be up for it, but in this case you can see that we use `props` as a whole twice on L76 and L85.

File: assets/js/blocks/product-query/inspector-controls/product-selector.tsx (hunk 3 above)
Comment (sunyatasattva) [reply assets/js/blocks/product-query/inspector-controls/product-selector.tsx]: Yea good call.

**Pull request merged**